### PR TITLE
Make `@ContributesTo` repeatable.

### PR DESCRIPTION
**Pull request by vRallev**
While you now can have multiple `@ContributesTo` annotations for Dagger modules and component interfaces, they all must have a different scope (contributing to the same scope twice is redundant).

This is a significant change and many of the internals of Anvil have to be updated.

There was a bug in Kotlin `1.6.0` [KT-49651](https://youtrack.jetbrains.com/issue/KT-49651) that has been fixed in `1.6.10`. I was able to remove the workarounds again. 

I apologize for the size of the PR, but there was simply no way to break this down. In fact, I had separate commits initially, but decided to squash them, because several tests were broken with the intermediate commits. I suggest looking for weird Kotlin idioms and looking at the tests. 

I don't plan to merge this PR soon. Probably next year when I'll try to make the other annotations repeatable as well.

### CODE REVIEW

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt [hunk 1]
```diff
@@ -91,62 +92,85 @@ public fun KtAnnotated.requireAnnotation(
 }
 
 @ExperimentalAnvilApi
+public fun KtAnnotated.requireAnnotations(
+  fqName: FqName,
+  module: ModuleDescriptor
+): List<KtAnnotationEntry> {
+  return findAnnotations(fqName, module)
+    .ifEmpty { null }
```
Comment (0xMatthewGroves) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:100]: takeIfNotEmpty()? You just created the extension.

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt [hunk 2]
```diff
@@ -91,62 +92,85 @@ public fun KtAnnotated.requireAnnotation(
...
   module: ModuleDescriptor
-): KtAnnotationEntry? {
+): KtAnnotationEntry? = findAnnotations(fqName, module).singleOrEmpty()
+
+@ExperimentalAnvilApi
+public fun KtAnnotated.findAnnotations(
+  fqName: FqName,
+  module: ModuleDescriptor
+): List<KtAnnotationEntry> {
   val annotationEntries = annotationEntries
```
Comment (0xMatthewGroves) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:119]: Doesn't look like it.

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt [hunk 3]
```diff
@@ -91,62 +92,85 @@ public fun KtAnnotated.requireAnnotation(
...
-  if (annotationEntries.isEmpty()) return null
+  if (annotationEntries.isEmpty()) return emptyList()
```
Comment (0xMatthewGroves) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:120]: Similar comment, but the logic below will return an emptyList() automatically if it's empty, so why do the initial check first? It's my understanding that the collection extensions have early bail out mechanisms baked in. i.e. filter {}

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt [hunk 4]
```diff
@@ -159,16 +183,57 @@ public fun KtClassOrObject.scope(
...
 
 @ExperimentalAnvilApi
+public fun KtClassOrObject.scopes(
+  annotationFqName: FqName,
+  module: ModuleDescriptor
+): List<FqName> {
+  return requireAnnotations(annotationFqName, module)
+    .map { it.scope(module) }
+    .also { scopes ->
+      // Exit early to avoid allocating additional collections.
```
Comment (0xMatthewGroves) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:216]: 👍

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt [hunk 5]
```diff
@@ -159,16 +183,57 @@ public fun KtClassOrObject.scope(
...
+            "${duplicates.keys.joinToString()}. Contributing multiple times to the same scope " +
+            "is forbidden and all scopes must be distinct.",
+          element = this
+        )
+      }
+    }
+}
+
+@ExperimentalAnvilApi
+// TODO Repeatable: deprecate and add other function.
```
Comment (0xMatthewGroves) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:236]: 🤔

File: compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 6]
```diff
@@ -46,6 +47,9 @@ public class AnvilCompilation internal constructor(
     enableExperimentalAnvilApis: Boolean = true,
   ) = apply {
     checkNotCompiled()
+    check(!anvilConfigured) { "Anvil should not be configured twice." }
```
Comment (0xMatthewGroves) [inline compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt:50]: Is this related to the @ContributesTo changes or should this be a separate commit/PR? Seems unrelated to me...

File: compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt [hunk 7]
```diff
@@ -1,98 +1,177 @@
...
         // Check that the annotation really is present. It should always be the case, but it's
         // a safetynet in case the generated properties are out of sync.
-        it.annotationOrNull(annotation, scope) != null
+        it.annotations(annotation, scope).isNotEmpty()
+      }
+  }
+
+  /**
+   * Returns a sequence of contributed hints from the dependency graph. Note that the result
+   * includes inner classes already.
```
Comment (0xMatthewGroves) [inline compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt:54]: Mega nit: Note that the result already includes inner classes.

File: compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt [hunk 8]
```diff
@@ -1,98 +1,177 @@
...
+            null
+          }
+        }
+        else -> null
+      }
+    }
   }
+}
+
+// TODO Repeatable: Cleanup this class with what's actually required.
```
Comment (0xMatthewGroves) [inline compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt:130]: Not sure if you use github to track your issues, but worthwhile to note your ticket number here if you do.

File: compiler/src/main/java/com/squareup/anvil/compiler/InterfaceMerger.kt [hunk 9]
```diff
@@ -36,45 +38,58 @@ internal class InterfaceMerger(
...
     if (!DescriptorUtils.isInterface(thisDescriptor)) {
       throw AnvilCompilationException(thisDescriptor, "Dagger components must be interfaces.")
     }
 
+    val module = thisDescriptor.module
+
+    val scope = try {
+      mergeAnnotation.scope(module)
+    } catch (e: AssertionError) {
+      // In some scenarios this error is thrown. Throw a new exception with a better explanation.
```
Comment (0xMatthewGroves) [inline compiler/src/main/java/com/squareup/anvil/compiler/InterfaceMerger.kt:66]: 👍

**Review by 0xMatthewGroves — APPROVED** [review]
Whew. Big PR 😅

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt (hunk 1 above)
Comment (vRallev) [reply compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:100]: For that I had to change the signature slightly. It's working now.
```kotlin
public inline fun <C, T> C.takeIfNotEmpty(): C? where C : Collection<T> = ifEmpty { null }
```

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt (hunk 2 above)
Comment (vRallev) [reply compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:119]: It's an improvement, because I don't want to call the getter function multiple times. That's unnecessary overhead, because it might create a new collection every time.

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt (hunk 3 above)
Comment (vRallev) [reply compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:120]: The operations below are quite heavy and returning early here happens quite often. It's a quick, faster exit.

File: compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt (hunk 6 above)
Comment (vRallev) [reply compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt:50]: It was necessary for one specific test that verifies contributions from the old format are picked up. This is just a sanity check.

File: compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt (hunk 7 above)
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt:54]: Done

File: compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt (hunk 8 above)
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt:130]: I have to clean up all deprecated methods and TODOs later. This PR would otherwise explode even further.

File: compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt [hunk 10]
```diff
@@ -618,4 +618,226 @@ class InterfaceMergerTest(
...
+      """
+    ) {
+      assertThat(componentInterface extends contributingInterface).isFalse()
+      assertThat(componentInterface extends secondContributingInterface).isTrue()
+      assertThat(subcomponentInterface extends contributingInterface).isTrue()
+      assertThat(subcomponentInterface extends secondContributingInterface).isFalse()
+    }
+  }
+
+  @Test fun `contributed interfaces in the old format are picked up`() {
```
Comment (JoelWilcox) [inline compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt:800]: What does the old format mean here?

File: compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt [hunk 11]
```diff
@@ -618,4 +618,226 @@ class InterfaceMergerTest(
...
+    val result = AnvilCompilation()
+      .useIR(USE_IR)
+      .enableAnvil(false)
```
Comment (JoelWilcox) [inline compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt:803]: Also curious if maybe `enableAnvil` should be called something else; at least, when I read this test and see that we are disabling Anvil for AnvilCompilation, but the test is for contributed interfaces being picked up, I would expect nothing to be picked up since Anvil was disabled. So perhaps this is a misunderstanding on my part

File: compiler/src/test/java/com/squareup/anvil/compiler/ModuleMergerTest.kt [hunk 12]
```diff
@@ -1119,6 +1110,244 @@ class ModuleMergerTest(
...
+    ) {
+      assertThat(componentInterface.anyDaggerComponent.modules.withoutAnvilModule())
+        .containsExactly(daggerModule2.kotlin)
+
+      assertThat(subcomponentInterface.anyDaggerComponent.modules.withoutAnvilModule())
+        .containsExactly(daggerModule1.kotlin)
+    }
+  }
+
+  @Test fun `contributed modules in the old format are picked up`() {
```
Comment (JoelWilcox) [inline compiler/src/test/java/com/squareup/anvil/compiler/ModuleMergerTest.kt:1305]: The ample test coverage in this PR is awesome 🔥

**Review by JoelWilcox — COMMENTED** [review]
Overall looks great! A number of my comments are related to me not having dug into the anvil implementation previously and just looking to better understand things :)

**Comment by vRallev** [comment]
Closing this PR in favor of #551.
